### ISSUE TITLE
Support npm module "feathers"

### ISSUE THREAD

**Issue by camstuart** [eclipsesource/tabris-js#1293]
Please add support for `feathers` npm modules and friends:

```
feathers
feathers-authentication-client
feathers-hooks
feathers-socketio
socket.io-client
```
This would allow for a very convenient and easy to use back end option for Tabris-js apps.

Feathersjs is a very thin layer on top of express that makes creating realtime API's simple and fast.

What interests me in particular is the ease at which I would be able to consume data stored in a back end application server, with realtime support baked in.

Currently adding any of these modules results in either:

`Cannot find module 'tty'`
or
`Cannot find module 'http'`

**Comment by camstuart** [comment]
Thanks to the help of @cookieguru , I managed to get a little bit further, doing:
`npm install ttys tabris-js-node --save` and in the top of my scripts:

```JavaScript
require('tabris-js-node');
require('ttys');
```

But then I hit:

`Cannot find module 'fs'` and wasn't able to get past it

**Comment by mpost** [comment]
Correct me if i am wrong but i think the `fs` module will not work on tabris, as it uses native parts to interface with the underlying os.

**Comment by cookieguru** [comment]
@mpost Would it depend on what you are accessing?  `require` hits the "disk".  But anything outside the Cordova file structure would probably need native code (and therefore the file permissions), a la the various Cordova file plugins. 

If the plugins is just caching data then a `fs` mock implementation that leverages `localStorage` might work.

**Comment by mpost** [comment]
@cookieguru Yeah that could be an option but the localStorage API is pretty limited compared to the `fs`module.

**Comment by camstuart** [comment]
I wonder though if it's truly required, or could be stubbed? (but I don't know how to go about that and need an example)

In the feathers client, specifically `feathers-authentication-client`, there is an option to store json web tokens in `localStorage` which I presume is the reason for the `fs` module.

When feathersjs is being used in their [react-native chat example](https://github.com/feathersjs/feathers-react-native-chat/blob/master/src/Store.js), the storage option `AsyncStorage` is used.

I imagine there is a way to write custom storage strategies, utilising Tabris local storage.
At least that is my hope anyway.

How can I fool the environment into thinking `fs` is present so I can continue to investigate if I can actually use the feathers modules? 

The local storage doesn't interest me, it's the web socket `socket.io` capability that attracts me, and the very convenient way of interacting with the backend. rest calls are becoming very tiresome.

See also: https://github.com/feathersjs/feathers/issues/571

**Comment by cookieguru** [comment]
> How can I fool the environment into thinking `fs` is present so I can continue to investigate if I can actually use the feathers modules?

You'd need to research how feathers is using `fs` and then go from there.  I see you've added a polyfill for that, so the next case would be to see how `stream` is being used.

**Comment by camstuart** [comment]
Indeed @cookieguru .... but I'm tipping we are getting closer to the heart of the problem, surely `stream` is needed for web socket support.

But hopefully later today I'll have some time to crack these modules open and take a peek

**Comment by ralfstx** [comment]
As for `fs`, we already discussed a native filesystem API, as the cordova plugins for file access are hard to use and limited. We could think about supporting a (small) subset of the Node.js fs API as well. It's more a mid-term goal and won't help you with this problem now, just wanted to let you know...
